src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -795,39 +817,53 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                         int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
                         if (readBytes == 0)
                         {
-                            return 0;
+                            receivedEOF = true;
```

Comment:
I believe this logic was wrong (and was wrong before).

We should only treat readBytes == 0 as EOF here if we don't have any bytes already buffered. 

This can only happen on the first read through this loop, **but** just the fact that it is the first read doesn't guarantee we don't have bytes already buffered from previous read calls.

I think this actually ends up making the logic simpler, since you don't actually need to treat the first read and subsequent reads differently. You just need to fail the case when we readBytes == 0 but we already have some data buffered.

Reply:
I don't think I fully understand the comment. I think there are few basic cases:

1. we got EOF when trying to read beginning of the frame and there are no pending data.  This is clear EOF and we simple return 0 back to the user. 
2. Same as above but we have decrypted data. We need to dispatch them first but we need to remember the EOF was already reached so it return 0 on next read. We can dispatch the data and we can try to read again instead remembering if we feel that would be safe and acceptable. It feels that would be more work but we do not need to remember state. 
3. We get EOF while trying to read rest of the frame. That may be viewed as error.

Reply:
I'm pointing out that we don't handle all the variations of (3) correctly today.

In particular, if there is encrypted data in the buffer on entry, and we do a read against the underlying stream and get EOF, we will not treat this as an error.

Reply:
ok. I can fix that. Or call underlying OS function and let it decide if that is failure or not.

Reply:
I moved the processing to helper function. It will throw if we get EOF in middle of TLS frame.